### PR TITLE
Task discovery cli baseline

### PR DESCRIPTION
**Pull request by Tomdye**
# Work towards issue: #10 

Will display help including generated task descriptions when calling
```
dojo
```

## Example display
screenshot 2016-08-18 13 54 50

## Code coverage
screen shot 2016-08-25 at 16 01 17

### CODE REVIEW

File: src/lib/index.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+function getCommandDescription(commandType: string, commandSubTypes: CommandConfig[]): string {
+	return commandSubTypes.length > 1 ?
+		`There are ${commandSubTypes.length} ${commandType} subCommands: ${commandSubTypes.map((commandSubType: CommandConfig) => commandSubType.name).join(', ')}` :
+		commandSubTypes[0].description;
+}
+
+function loadCommandFromPath(path: string): void {
+	const { description, register, run } = require(path);
+	const pluginParts = /dojo-cli-(.*)-(.*)/.exec(path);
```
Comment (matt-gadd) [inline src/lib/index.ts]: this should come from the `searchPrefixes` in the config no?

Comment (Tomdye) [inline src/lib/index.ts]: @matt-gadd would it not make more sense to remove `prefixes` from the config / make it a `string` rather than a `string[]` to be more opinionated about the `dojo-cli-` command naming conventions?

Comment (matt-gadd) [inline src/lib/index.ts]: @Tomdye i'd still leave it in the config (as it's used for the folder discovery at the moment), but maybe just change it to a string rather than an array of strings.

File: src/lib/config.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	searchPrefixes: string[]
+};
+
+export default <Config> {
+	searchPaths: [
+		resolve(__dirname, '../node_modules'),
+		resolve(__dirname, '../../'),
+		'node_modules'
+	],
+	searchPrefixes: ['dojo-cli']
```
Comment (Tomdye) [inline src/lib/config.ts]: I think this should become `searchPrefix: 'dojo-cli'`

File: src/util/setupLogger.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { add, remove, transports } from 'winston';
+
+export interface VerboseOptions {
+	verbose?: boolean;
+}
+
+export default function setupLogger(verbose: boolean = false) {
```
Comment (matt-gadd) [inline src/util/setupLogger.ts]: do we need this?

File: src/util/setupLogger.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	remove(transports.Console);
+	add(transports.Console, {
+		showLevel: false,
+		level: verbose ? 'verbose' : 'info'
+	});
+
+	add(transports.File, {
+		filename: '.dojo-cli.log',
```
Comment (matt-gadd) [inline src/util/setupLogger.ts]: i think we discussed this before, but this is a bad place to put the .log, especially in places where your commands demand an empty folder.

File: src/util/setupUpdateNotifier.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import * as updateNotifier from 'update-notifier';
```
Comment (matt-gadd) [inline src/util/setupUpdateNotifier.ts]: can this file just be named updateNotifier? I don't know why we need to prefix with `setup` everywhere.

File: src/util/setupLogger.ts (hunk 4 above)
Comment (agubler) [inline src/util/setupLogger.ts]: Do we even want to have a `File` transport?

Comment (matt-gadd) [inline src/util/setupLogger.ts]: I'm not even sure, I don't think this is in mvp anyway?

File: src/lib/index.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+const helpEpilog = `For more information on any of these commands just run them with '-h'.
+
+e.g. 'dojo run -h' will give you the help for the 'run' command.
+
+(You are running dojo-cli ${pkg.version})`;
+
+function getCommandDescription(commandType: string, commandSubTypes: CommandConfig[]): string {
+	return commandSubTypes.length > 1 ?
+		`There are ${commandSubTypes.length} ${commandType} subCommands: ${commandSubTypes.map((commandSubType: CommandConfig) => commandSubType.name).join(', ')}` :
```
Comment (matt-gadd) [inline src/lib/index.ts]: we don't use camel case in plain english :P. how about just `commands` anyway?

File: src/util/setupLogger.ts (hunk 4 above)
Comment (Tomdye) [inline src/util/setupLogger.ts]: i'll take the file transport out then, users can use the verbose flag to see verbose output

**Comment by matt-gadd** [comment]
it feels like the file structure to this is kind of just split up arbitrarily from the poc? I think it could be better organises logically, at the moment we seem to have files which serve one function and are named after it, yet others aren't?

File: src/util/setupLogger.ts (hunk 3 above)
Comment (Tomdye) [inline src/util/setupLogger.ts]: do we need to be able to enable verbose logging? Whats wrong with it?

File: src/lib/index.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		commandSubTypes[0].description;
+}
+
+const commandRegExp = new RegExp(`${config.searchPrefix}-(.*)-(.*)`);
+function loadCommandFromPath(path: string): void {
```
Comment (matt-gadd) [inline src/lib/index.ts]: given all these functions have the word `command` in, perhaps create a `command` module and these can then all be nicely named `load`, `register`, `getDescription` etc

Comment (Tomdye) [inline src/lib/index.ts]: I thought about that, they'd all need to be passed the `taskMap` and `taskSet` though

Comment (matt-gadd) [inline src/lib/index.ts]: maybe evaluate how you construct this then.

File: src/util/setupLogger.ts (hunk 3 above)
Comment (matt-gadd) [inline src/util/setupLogger.ts]: like how useful is the logging in the mvp?

Comment (Tomdye) [inline src/util/setupLogger.ts]: it's already there and it's small so I don't see the problem

Comment (matt-gadd) [inline src/util/setupLogger.ts]: okay, well shall we remove the File transport and remove the `setup` from the filename

File: src/lib/index.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import setupLogger, { VerboseOptions } from '../util/setupLogger';
+import setupUpdateNotifier from '../util/setupUpdateNotifier';
+import getGlobPaths from '../util/getGlobPaths';
+import * as chalk from 'chalk';
+import * as globby from 'globby';
+import config from './config';
+import { log } from 'winston';
+const pkg = require('../../package.json');
+
+type CommandConfig = {
```
Comment (matt-gadd) [inline src/lib/index.ts]: we probably want this in a separate file as it's the interface for consumers.

**Comment by Tomdye** [comment]
@matt-gadd reopened with re-structuring

File: tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
+{
+	"version": "1.8.10",
```
Comment (agubler) [inline tsconfig.json]: `2.0.0`?

File: typings.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (agubler) [inline typings.json]: doesn't the template have a `name` property?

File: tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+	"compilerOptions": {
+		"declaration": false,
+		"module": "commonjs",
+		"noImplicitAny": true,
+		"outDir": "_build/",
+		"removeComments": false,
+		"sourceMap": true,
+		"target": "es6"
```
Comment (agubler) [inline tsconfig.json]: add `noImplicitThis` and `strictNullChecks`

File: src/logger.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { add, remove, transports, log } from 'winston';
+
+export interface VerboseOptions {
+	verbose?: boolean;
+}
+
+export default function setupLogger(verbose: boolean = false) {
+	remove(transports.Console);
```
Comment (agubler) [inline src/logger.ts]: How come you have to remove the `Console` transport first?

File: src/index.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+globby(globPaths(config)).then((paths) => {
+	verbose(`index - loading commands`);
+	const commandSet: command.CommandSet = new Set();
+	paths.forEach((path) => {
+		const commandConfig = command.load(path, commandSet);
+		const commandType = commandConfig.type;
+
+		if (commandsMap.has(commandType)) {
+			commandsMap.get(commandType).push(commandConfig);
```
Comment (matt-gadd) [inline src/index.ts]: feels a bit side effecty to just do a `get` with a `push` rather than a `set` here? would this block not read nicer as:

```js
const commands = commandsMap.get(commandType) || [];
commandsMap.set(commandType, [...commands, commandConfig]);
```

File: src/command.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+export function getDescription(commandType: string, commandSubTypes: CommandConfig[]): string {
+	return commandSubTypes.length > 1 ?
+		`There are ${commandSubTypes.length} ${commandType} commands: ${commandSubTypes.map((commandSubType: CommandConfig) => commandSubType.name).join(', ')}` :
+		commandSubTypes[0].description;
+}
+
+export function register(commandsMap: CommandsMap): void {
+	for (let [ commandType, commandSubTypes ] of commandsMap.entries()) {
+		verbose(`index:registerCommand - commandType: ${commandType}, commandSubTypes: ${JSON.stringify(commandSubTypes)}`);
```
Comment (matt-gadd) [inline src/command.ts]: have you thought about the structure of these logs? seems pretty verbose and the majority of it could be deduced rather than having these long manual templates in each function?

File: src/logger.ts (hunk 12 above)
Comment (Tomdye) [inline src/logger.ts]: because it seems to be the only way to override it, actually, i think i can perhaps improve this now that `logger` provides the `log` functions as we can have our own logger rather than the global `winston` one

File: src/command.ts [hunk 15]
```diff
@@ -0,0 +1,62 @@
+import { verbose } from './logger';
+import config from './config';
+import { command } from 'yargs';
+
+export interface Command {
```
Comment (matt-gadd) [inline src/command.ts]: should this be in a separate file as it's an interface a consumer may use?

File: src/command.ts [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+	name: string;
+	description: string;
+	register: Function;
```
Comment (matt-gadd) [inline src/command.ts]: typings for these?

File: src/command.ts [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+export interface CommandConfig extends Command {
+	type: string;
+	subType: string;
+}
+
+export type CommandsMap = Map<string, CommandConfig[]>;
+
+export type CommandSet = Set<string>;
```
Comment (agubler) [inline src/command.ts]: Do we need `CommandSet` and `CommandMap`?

File: src/command.ts (hunk 15 above)
Comment (Tomdye) [inline src/command.ts]: can't they use it from here?

File: src/command.ts (hunk 17 above)
Comment (Tomdye) [inline src/command.ts]: was going to look into that

File: src/command.ts (hunk 15 above)
Comment (matt-gadd) [inline src/command.ts]: i thought for external interfaces we preferred them in their own interfaces file?

Comment (agubler) [inline src/command.ts]: We will actually create a `d.ts` file for all the modules with the module typings.

Comment (agubler) [inline src/command.ts]: Misunderstood, it will still contain module interface as well as the public types/interfaces so doesn't change @matt-gadd's comment

Comment (matt-gadd) [inline src/command.ts]: @agubler i get that but this is actually an external interface we want people to consume. we don't actually want them to consume this module. something like: https://github.com/dojo/routing/blob/master/src/interfaces.d.ts

File: src/logger.ts (hunk 12 above)
Comment (agubler) [inline src/logger.ts]: Can you just do 

```js
  logger.configure({
    level: 'verbose',
    showLevel: false
  });
```

File: src/command.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+	};
+}
+
+export function getGroupDescription(group: string, commands: CommandsMap): string {
+	const commandNames: string[] = Array.from(commands.keys());
+	const numCommands = commandNames.length;
+	if (numCommands > 1) {
+		return `There are ${numCommands} ${group} commands: ${commandNames.join(', ')}`;
+	} else {
+		let { description } = <CommandWrapper> commands.get(commandNames[0]);
```
Comment (agubler) [inline src/command.ts]: `let`?

File: tests/unit/all.ts [hunk 19]
```diff
@@ -1 +1,3 @@
+import * as winston from 'winston';
```
Comment (matt-gadd) [inline tests/unit/all.ts]: you can drop this now right?

File: src/interfaces.ts [hunk 20]
```diff
@@ -0,0 +1,23 @@
+import { Argv, Yargs } from 'yargs';
+
+export interface RunResult {
+	success: boolean;
+	payload: any;
+}
+
+export interface TaskHelper {
```
Comment (matt-gadd) [inline src/interfaces.ts]: needs renaming. are we going to implement these in this PR?

File: src/CommandHelper.ts [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+		this.groupsMap = groupsMap;
+	};
+	groupsMap: GroupsMap;
+	run(group: string, taskName?: string, args?: yargs.Argv): Promise<RunResult> {
+		const command = getCommand(this.groupsMap, group, taskName);
+		if (command) {
+			return command.run(new Helper(this, yargs, {}));
+		}
+		else {
+			return Promise.resolve({
```
Comment (matt-gadd) [inline src/CommandHelper.ts]: we should probably `reject` the Promise, we can probably drop the `success` from the `CommandResult` also.

File: src/command.ts [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+	}
+}
+
+function pad(str: string, num: number): string {
+	const padding = new Array(num - str.length).join(' ');
+	return str + padding;
+}
+
+function getMultiCommandDescription(commands: CommandsMap): string {
+	const commandNames = Array.from(commands.keys(), (name) => name.length);
```
Comment (matt-gadd) [inline src/command.ts]: `commandNameLengths`

File: src/CommandHelper.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+	const commandKey = commandName ? `${group}-${commandName}` : group;
+	return commandsMap.get(commandKey);
+}
+
+export default class implements CommandHelper {
+	constructor(commandsMap: CommandsMap, context: any) {
+		this.commandsMap = commandsMap;
+		this.context = context;
+	};
+	commandsMap: CommandsMap;
```
Comment (matt-gadd) [inline src/CommandHelper.ts]: should the `commandsMap` be on here? doesn't seem like a thing we should expose given the `exists` and `run` apis?

File: src/index.ts [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+import config from './config';
+import loadCommands from './loadCommands';
+import registerCommands from './registerCommands';
+import { load, setSearchPrefix } from './command';
+const pkg = <any> require('../../package.json');
+
+async function init() {
+	updateNotifier(pkg, 0);
+	setSearchPrefix(config.searchPrefix);
+	const { commandsMap, yargsCommandNames } = await loadCommands(yargs, config, load);
```
Comment (agubler) [inline src/index.ts]: why do you need to pass `load` to this function? Can it nor just import the `load` function?

File: src/loadCommands.ts [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+		globby(globPaths, (<globby.Options> { ignore: '**/*.map' })).then((paths) => {
+			const commandsMap: CommandsMap = new Map();
+			const yargsCommandNames: YargsCommandNames = {};
+
+			paths.forEach((path) => {
+				const commandWrapper = load(path);
+				const { group, name } = commandWrapper;
+				const compositeKey = `${group}-${name}`;
+
+				// First of each type will be 'default' for now
```
Comment (matt-gadd) [inline src/loadCommands.ts]: perhaps make this code a functions? then we at least have the concept of a `getDefaultCommand` to later refactor on?

File: src/index.ts [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+import updateNotifier from './updateNotifier';
+import config from './config';
+import loadCommands from './loadCommands';
+import registerCommands from './registerCommands';
+import { load, setSearchPrefix } from './command';
+const pkg = <any> require('../../package.json');
+
+async function init() {
+	updateNotifier(pkg, 0);
+	setSearchPrefix(config.searchPrefix);
```
Comment (agubler) [inline src/index.ts]: For some reason I don't like this random function to set config as a global in another class - Would it be neater if the command class exported a function that you called to "init" and return the load function? 

so like (and you wouldn't need to do it in here)

```js
import initCommandLoader from './command';
const commandLoader = initCommentLoader(config);
```

Using `commandLoader` just like `load` now

File: src/interfaces.ts [hunk 27]
```diff
@@ -0,0 +1,22 @@
+import { Argv, Yargs } from 'yargs';
+
+export interface RunResult {
```
Comment (agubler) [inline src/interfaces.ts]: Did we decided to leave the wrapper `RunResult` for the `payload` still?

File: tests/unit/command.ts [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+			assert.equal(testName, commandWrapper.name);
+		},
+		'Should get description from loaded file'() {
+			assert.equal('test-description', commandWrapper.description);
+		},
+		'Should get register function from loaded file'() {
+			assert.isTrue(typeof commandWrapper.register === 'function');
+		},
+		'Should get run function from loaded file'() {
+			assert.isTrue(typeof commandWrapper.run === 'function');
```
Comment (matt-gadd) [inline tests/unit/command.ts]: we should be able to check the equality of this against the real function.

**Comment by agubler** [comment]
I generally think some more "doc" would be super useful

File: src/interfaces.ts (hunk 27 above)
Comment (Tomdye) [inline src/interfaces.ts]: might need removing

File: src/index.ts (hunk 24 above)
Comment (Tomdye) [inline src/index.ts]: unable to stub it for testing otherwise

File: src/CommandHelper.ts (hunk 23 above)
Comment (Tomdye) [inline src/CommandHelper.ts]: will mark as `private`

File: tests/unit/command.ts (hunk 28 above)
Comment (Tomdye) [inline tests/unit/command.ts]: cheers

File: src/index.ts (hunk 24 above)
Comment (agubler) [inline src/index.ts]: :trollface:

File: src/index.ts (hunk 26 above)
Comment (Tomdye) [inline src/index.ts]: sounds fair, so more of a factory

**Comment by Tomdye** [comment]
@ant will add doc

**Comment by agubler** [comment]
👍
